Modules/_functoolsmodule.c

```diff
@@ -560,37 +560,25 @@ functools_reduce(PyObject *self, PyObject *args)
         return NULL;
     }
 
-    if ((args = PyTuple_New(2)) == NULL)
-        goto Fail;
-
     for (;;) {
-        PyObject *op2;
-
-        if (args->ob_refcnt > 1) {
```

Comment:
I expect a ltitle slowdown if the function doesn't support fastcall. For example, a Python object with a __call__() method.

Can you please try to run a quick benchmark on the following two cases:
* function supporting FASTCALL, like a builtin function
* function not supporting FASTCALL

In general, I'm in favor of removing such hack, since playing with reference count can lead to complex and annoying bugs. But I would prefer to first see the cost on performances.